paccmann_chemistry/models/vae.py

```diff
@@ -182,18 +313,80 @@ def decoder_train_step(self, latent_z, input_seq, target_seq):
             The cross-entropy training loss for the decoder.
         """
         hidden = self.latent_to_hidden(latent_z)
-        stack = self.init_stack()
+        stack = self.init_stack
+
+        loss = self._forward_fn(input_seq, target_seq, hidden, stack)
+        return loss
+
+    def _forward_pass_padded(self, input_seq, target_seq, hidden, stack):
+        """The Decoder Train Step.
+
+        Note: Input and target sequences are outputs of
+            sequential_data_preparation(batch) with batches returned by
+            a DataLoader object.
+        Returns:
+            The cross-entropy training loss for the decoder.
+        """
+        if isinstance(input_seq, nn.utils.rnn.PackedSequence) and \
+                not isinstance(input_seq, torch.Tensor):
+            raise TypeError('Input is PackedSequence or is not a Tensor')
+
         loss = 0
         outputs = []
         for input_entry, target_entry in zip(input_seq, target_seq):
             output, hidden, stack = self(
                 input_entry.unsqueeze(0), hidden, stack
             )
+            output = self.output_layer(output).squeeze()
             loss += self.criterion(output, target_entry.squeeze())
             outputs.append(output)
 
         # For monitoring purposes
-        outputs = torch.argmax(torch.stack(outputs, -1), 1)
+        self.outputs = torch.argmax(torch.stack(outputs, -1), 1)
+
+        return loss
+
+    def _forward_pass_packed(self, input_seq, target_seq, hidden, stack):
+        """The Decoder Train Step.
+
+        Note: Input and target sequences are outputs of
+            sequential_data_preparation(batch) with batches returned by a
+            DataLoader object.
+        Returns:
+            The cross-entropy training loss for the decoder.
+        """
+        if not isinstance(input_seq, nn.utils.rnn.PackedSequence):
+            raise TypeError('Input is not PackedSequence')
+
+        loss = 0
+        input_seq_packed, batch_sizes = utils.perpare_packed_input(input_seq)
+        # Target sequence should have same batch_sizes as input_seq
+        input_seq_packed, _ = utils.perpare_packed_input(target_seq)
+        prev_batch = batch_sizes[0]
+
+        outputs = []
```

Comment:
Do we need to store the outputs?

Reply:
nope, only in the `_forward_pass_padded` so that we can look at an example reconstruction (for `_forward_pass_packed` that's a bit trickier, so I omit it for the moment)